templates/tftmpl/tmplfunc/service_regex.go

```diff
@@ -0,0 +1,258 @@
+package tmplfunc
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	consulapi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-bexpr"
+	"github.com/hashicorp/hcat"
+	"github.com/hashicorp/hcat/dep"
+	"github.com/pkg/errors"
+)
+
+var (
+	// queryParamOptRe is the regular expression to distinguish between query
+	// params and filters, excluding the regex parameter. Non-regex query parameters
+	// only have one "=" where as filters can have "==" or "!=" operators.
+	queryParamOptRe = regexp.MustCompile(`[\w\d\s]=[\w\d\s]`)
+)
+
+// serviceRegexFunc returns information on registered Consul
+// services that have a name that match a given regex. It queries
+// the Catalog List Services API initially to get all the services
+// and then queries the Health API for each matching service.
+// It supports parameters filter, dc, ns, and node-meta on the
+// Health API query only.
+//
+// Endpoints:
+//   /v1/catalog/services
+//   /v1/health/service/:service
+// Template: {{ serviceRegex regexp=<regex> <options> ... }}
+func serviceRegexFunc(recall hcat.Recaller) interface{} {
+	return func(opts ...string) ([]*dep.HealthService, error) {
+		result := []*dep.HealthService{}
+
+		d, err := newServiceRegexQuery(opts)
+		if err != nil {
+			return nil, err
+		}
+
+		if value, ok := recall(d); ok {
+			return value.([]*dep.HealthService), nil
+		}
+
+		return result, nil
+	}
+}
+
+// serviceRegexQuery is the representation of the regex service
+// query from inside a template.
+type serviceRegexQuery struct {
```

Comment:
That makes sense to me! I originally went with the singular service because I was thinking of how hcat has service for health service and services for catalog services. But now that you've pointed it out, I realize that it's returning multiple services (unlike the `service` method) so plural makes more sense.